### ISSUE TITLE
Config and buffering issues on Vizio & Toshiba

### ISSUE THREAD

**Issue by vitalist82** [google/shaka-player#1473]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**: YES

**What version of Shaka Player are you using?**: 2.3.6

**Can you reproduce the issue with our latest release version?**: YES

**Can you reproduce the issue with the latest code from `master`?**: YES

**Are you using the demo app or your own custom app?**: custom app (demo app is not playing our streams)

**If custom app, can you reproduce the issue using our demo app?**: -

**What browser and OS are you using?**: VIZIO M50-E1 TV - Chromium-based, Toshiba (Netfront)

**What are the manifest and license server URIs?**:
Manifest URI: https://dc1vodcadazn-a.akamaihd.net/out/u/DC1_1p99262p0ner41x61o9crhlo77_DAZN_ICC_WES_SRI_T2_D4_FUL_170618_1529275743650-encrypted.mpd

License server URI is only valid for limited amount of time, I can send it by e-mail.

**What did you do?**

After starting playback of the stream mentioned above, player seems to get stuck in `onPollWindow_` in `playhead.js` because of a truncating issue.

`currentTime` is equal to `0.061333`, while `seekStart` is `0.06133333333333333`, thus the condition `if (currentTime < seekStart)` is evaluated as `true`.

Because of this `this.video_.currentTime` is reassigned to `0.06133333333333333`, but after this assignment, the value remains `0.061333`.

After few calls of `onPollWindow_` playback fails.

When the condition is changed to something like `if (currentTime < (seekStart - 0.1))`, playback is working.

I don't have the overall picture and don't know, how the startTime value is calculated yet.

Is this something, that could be explained based on provided V2 logs?

[no-workaround-failed.txt](https://github.com/google/shaka-player/files/2133477/no-workaround-failed.txt)
[workaround-ok.txt](https://github.com/google/shaka-player/files/2133478/workaround-ok.txt)

Thank you

Tomas Beranek

**Comment by theodab** [comment]
So, if I'm understanding correctly, in the error scenario we seek forward, but since the video's currentTime value is incorrect (perhaps due to a bug in the NetFront browser's implementation of HTMLVideoElement), we think that the video hasn't seeked forward correctly, so we keep seeking?

If we were seeking endlessly through onPollWindow_, though, I would expect the v2 log to be full of messages like `Jumping forward ___ seconds to catch up with the seek range.`, but I didn't see any in the samples you provided.

Could you go into `lib/media/video_wrapper.js`, find the `movePlayhead_` function, and remove the contents of the function starting at line 249? We have a workaround for an IE bug there, but it sounds like that workaround could possibly be the cause of your problem. It makes us repeatedly seek if we detect that a seek has failed... for instance, if the HTMLVideoElement is reporting that the time hasn't changed to the desired value, which sounds like the case here. In #1471, we are considering investigating if that workaround is still necessary; right now that issue is marked for v2.6, but if getting rid of it fixes your problem, then I can definitely move it up to v2.5.

**Comment by vitalist82** [comment]
I did some further testing today and found out, that on VIZIO Smartcast TV, it is possible to play the stream without any workaround, just by providing a proper config to the player, although I don't understand, how exactly it fixes the issue.

Unfortunately, the same player configuration doesn't work on the other Netfront-based Toshiba.

I am setting following properties of the config (we are using the same settings for our Chromecast receiver app):

```
var restrictions:Restrictions = {
	minBandwidth: 0,
	maxBandwidth: maxBitrate
};

var config:PlayerConfiguration = {
	abr: {
		bandwidthDowngradeTarget: BANDWIDTH_DOWNGRADE_TARGET,
		bandwidthUpgradeTarget: BANDWIDTH_UPGRADE_TARGET,
		defaultBandwidthEstimate: startBitrate,
		restrictions: restrictions
	},
	streaming: {
		retryParameters: {
			timeout: STREAMING_TIMEOUT,
			maxAttempts: 3
		},
		rebufferingGoal: 1,
		bufferingGoal: 1,
		bufferBehind: 1
	},
	restrictions: restrictions
};
```

The workaround from `movePlayhead_ ` function hasn't been called during my tests, so it's probably not causing any issues on VIZIO Smartcast.

Debugging on Toshiba device is a bit more difficult than on VIZIO Smartcast and I need to get some logs in order to understand what's going on.

**Comment by joeyparrish** [comment]
We've never done any testing on Netfront, but that appears to be WebKit-based.  Of all the major browser implementations, Safari/WebKit gives us the most difficulty and seems to have the buggiest implementation of MediaSource.

Although Netfront is not on our list of officially-supported platforms, we will do what we can to help.  We will not be able to provide any help debugging a platform we don't have, but we will try to advise you as you work through the issue, and we would be happy to review a pull request for any workaround you come up with.

It is often the case that fudge factors are needed to account for inaccuracies in timestamps in various parts of the system, so your workaround seems reasonable on the surface.  Does your "- 0.1" workaround work on all of your TV platforms?

**Comment by vitalist82** [comment]
We already had some issues with the mentioned Toshiba TV, that we solved partly with your help and partly with help of the manufacturer. VIZIO SmartCast is a new device in our family.. Current issue is related to the 12s fragment size and should be resolved on the side of our stream provider by providing streams with lower fragment size in the future.

To sum up my latest test results:

After creating the config change for VIZIO SmartCast I tried removing as many as possible parts of the config to get the smallest set of settings that are necessary to enable playback and it turned out, that 'defaultBandwidthEstimate' is the most important one. I also increased the timeout of the application which is limiting the interval between setting the source and start of the playback. Previously it was set to 10s, which was sometimes too short to download all the 12s long segments in higher quality on a slower network.

After doing these changes and being happy with results, I provided the code to a team based on a different place, connected through a different, even slower VPN (also EU to US) and they observed some issues (playback was working only in few cases from many attempts). Later during the day, probably because the VPN connection got slower, I also started noticing issues with the same configuration, which led me to decreasing the defaultBandwidthEstimate to a lower bitrate, which worked for that moment.

From all the testing so far, it seems, that the main issue is the network connection being too slow for the 12s segments in higher bitrate. As next step, I'll check, whether the -0.1s workaround can improve the experience in some cases and I'll also increase the application's timeouts, so that the player is not limited by the time it needs to download all the segments before starting the playback.

I will update you if I'll find something new. Many thanks for your support.
